src/policy/policy-inl.h

```diff
@@ -0,0 +1,196 @@
+#ifndef SRC_POLICY_POLICY_INL_H_
+#define SRC_POLICY_POLICY_INL_H_
+
+#include "policy/policy.h"
+#include "node_errors.h"
+#include "env-inl.h"
+
+#include <array>
+#include <string>
+#include <vector>
+
+namespace node {
+namespace policy {
+
+Policy* Policy::GetCurrent(Environment* env) {
+  return GetCurrent(env->context());
+}
+
+Policy* Policy::GetCurrent(v8::Isolate* isolate) {
+  return GetCurrent(isolate->GetCurrentContext());
+}
+
+Policy* Policy::GetCurrent(v8::Local<v8::Context> context) {
+  if (UNLIKELY(context.IsEmpty())) {
+    return nullptr;
+  }
+  if (UNLIKELY(context->GetNumberOfEmbedderDataFields() <=
+               ContextEmbedderIndex::kPolicyIndex)) {
+    return nullptr;
+  }
+  Policy* policy = static_cast<Policy*>(
+      context->GetAlignedPointerFromEmbedderData(
+          ContextEmbedderIndex::kPolicyIndex));
+  return policy;
+}
+
+bool Policy::test(Permissions permission) const {
+  return !permissions_.test(static_cast<size_t>(permission));
+}
+
+#define V(name, _, parent)                                                     \
+  if (permission == Permissions::k##parent)                                    \
+    SetRecursively(Permissions::k##name, value);
+void Policy::SetRecursively(Permissions permission, bool value) {
+  if (permission != Permissions::kPermissionsRoot)
+    permissions_[static_cast<size_t>(permission)] = value;
+  PERMISSIONS(V)
+}
+#undef V
+
+template <typename...P>
+bool Policy::granted(P...permissions) const {
+  std::array<Permissions, sizeof...(permissions)> perms {
+    std::forward<P>(permissions)...
+  };
+  for (Permissions permission : perms) {
+    if (UNLIKELY(!test(permission)))
+      return false;
+  }
+  return true;
+}
```

Comment:
Some compilers seem to instantiate the `std::array` + roll up the loop when `permissions` large-ish (> 3 elements) … I don’t know if it matters to you, but I think we want to keep perf impact as minimal as possible, and I think it makes sense to address that? The classical approach to that would be recursion, i.e.

```c++
bool Policy::granted() const { return true; }

template <typename...P>
bool Policy::granted(Permission p, P...permissions) const {
  return LIKELY(test(p)) && granted(permissions...);
}
```

(and it might also make sense to move the `(UN)LIKELY()` into `test()` itself, I think the compiler should be clever enough to understand that)

Reply:
Yeah good idea on the recursion. To be honest, tho, I'm not sure how much I like the current pattern here. I like the type safe variadic templates using the enum class but I also just really prefer xor'd regular enums ;-). We can play around with this a bit to see what pattern works best in the end.